core/src/main/java/io/camunda/migrator/Pagination.java

```diff
@@ -88,25 +96,58 @@ public List<T> toList() {
     return list;
   }
 
+  public Map<String, Map<String, Object>> toVariableMap() {
+    Map<String, Map<String, Object>> result = new HashMap<>();
+    processVariables((variable, typedValue) -> {
+      Map<String, Object> variableResult = getVariableResult(variable, typedValue);
+      String activityInstanceId = variable.getActivityInstanceId();
+      Map<String, Object> variableMap = result.getOrDefault(activityInstanceId, new HashMap<>());
+      variableMap.putAll(variableResult);
+      result.put(activityInstanceId, variableMap);
+    });
+    return result;
+  }
+
+
+  public Map<String, Object> toVariableMapSingleActivity() {
+    Map<String, Object> variableResult = new HashMap<>();
+    processVariables((variable, typedValue) -> variableResult.putAll(getVariableResult(variable, typedValue)));
+    return variableResult;
+  }
+
   /**
    * Heads-up: this implementation needs to be null safe for the variable value.
    * Using streams might lead to undesired {@link NullPointerException}s.
    */
-  public Map<String, Object> toVariableMap() {
-    Map<String, Object> result = new HashMap<>();
+  private void processVariables(BiConsumer<VariableInstanceEntity, TypedValue> consumer) {
+    List<VariableInterceptor> interceptors = (List<VariableInterceptor>) context.getBeansOfType(
+        VariableInterceptor.class).values().stream().toList();
     toList().forEach(e -> {
-      VariableInstanceEntity var = (VariableInstanceEntity) e;
-      TypedValue typedValue = var.getTypedValue(false);
-      if (typedValue.getType().equals(ValueType.OBJECT)) {
-        // skip the value deserialization
-        result.put(var.getName(), typedValue.getValue());
-      } else if (typedValue.getType().equals(SpinValueType.JSON) || typedValue.getType().equals(SpinValueType.XML)) {
-        // For Spin JSON/XML, explicitly set the string value
-        result.put(var.getName(), typedValue.getValue().toString());
-      } else {
-        result.put(var.getName(), var.getValue());
+      VariableInvocation variableInvocation = new VariableInvocation((VariableInstanceEntity) e);
+      if (interceptors != null && !interceptors.isEmpty()) {
+        interceptors.forEach(i -> {
+          try {
+            i.execute(variableInvocation);
+          } catch (Exception ex) {
+            throw new VariableInterceptorException("An error occurred during variable transformation.", ex);
```

Comment:
💭 Maybe we could mention in the README that: 
if an interceptors throws an error no other interceptors will be called for that process instance. Mention that right now we don't give a tool to specify interceptor order.